gostream/codec/h264/ffmpeg/avcodec/avcodec.go

```diff
@@ -0,0 +1,272 @@
+// Package avcodec is a thin wrapper around FFmpeg/release4.4.
+// See: https://github.com/FFmpeg/FFmpeg/tree/release/4.4
+package avcodec
+
+//#cgo CFLAGS: -Wno-deprecated-declarations -I${SRCDIR}/../ffmpeg/include
+//#cgo LDFLAGS: -L${SRCDIR}/../ffmpeg/lib -lavformat -lavcodec -lavutil -lavdevice -lavfilter -lswresample -lswscale -lm -lX11
+//#include <libavformat/avformat.h>
+//#include <libavcodec/avcodec.h>
+//#include <libavcodec/packet.h>
+//#include <libavutil/avutil.h>
+import "C"
+import "unsafe"
+
+// AvPixFmtYuv420p the pixel format AV_PIX_FMT_YUV420P
+const AvPixFmtYuv420p = C.AV_PIX_FMT_YUV420P
+
+type (
+	// Codec an AVCodec
+	Codec C.struct_AVCodec
+	// Context an AVCodecContext
+	Context C.struct_AVCodecContext
+	// Dictionary an AVDictionary
+	Dictionary C.struct_AVDictionary
+	// Frame an AVFrame
+	Frame C.struct_AVFrame
+	// Packet an AVPacket
+	Packet C.struct_AVPacket
+	// PixelFormat an AVPixelFormat
+	PixelFormat C.enum_AVPixelFormat
+)
+
+// FindEncoderByName Find a registered encoder with the specified name.
+//
+// @param name the name of the requested encoder
+// @return An encoder if one was found, NULL otherwise.
+func FindEncoderByName(c string) *Codec {
+	return (*Codec)(C.avcodec_find_encoder_by_name(C.CString(c)))
+}
+
+// PacketAlloc Allocate an AVPacket and set its fields to default values.  The resulting
+// struct must be freed using av_packet_free().
+//
+// @return An AVPacket filled with default values or NULL on failure.
+//
+// @note this only allocates the AVPacket itself, not the data buffers. Those
+// must be allocated through other means such as av_new_packet.
+//
+// @see av_new_packet
+func PacketAlloc() *Packet {
+	return (*Packet)(C.av_packet_alloc())
+}
+
+// AllocContext3 Allocate an AVCodecContext and set its fields to default values. The
+// resulting struct should be freed with avcodec_free_context().
+//
+// @param codec if non-NULL, allocate private data and initialize defaults
+//
+//	for the given codec. It is illegal to then call avcodec_open2()
+//	with a different codec.
+//	If NULL, then the codec-specific defaults won't be initialized,
+//	which may result in suboptimal default settings (this is
+//	important mainly for encoders, e.g. libx264).
+//
+// @return An AVCodecContext filled with default values or NULL on failure.
+func (c *Codec) AllocContext3() *Context {
+	return (*Context)(C.avcodec_alloc_context3((*C.struct_AVCodec)(c)))
+}
+
+// SetEncodeParams sets the context's width, height, pixel format (pxlFmt), if it has b-frames and GOP size.
+func (ctxt *Context) SetEncodeParams(width, height int, pxlFmt PixelFormat, hasBFrames bool, gopSize int) {
+	ctxt.width = C.int(width)
+	ctxt.height = C.int(height)
+	ctxt.bit_rate = 200000
```

Comment:
Is this 200 Kbps? Seems reasonable for small image sizes but wondering if it is enough for 1080p+ frame sizes. How is the video quality for larger frames? Maybe we can scale bit_rate by size if necessary.

Reply:
This is a good point. I'll create a follow up tickets to see how these parameters impact performance but 10x this gave us as clear of an image we need right now for large images. For context, I'm using an E-Meet with a res of 1920x180.